src/librustc_ast/ast.rs

```diff
@@ -215,11 +215,18 @@ impl GenericArg {
 pub struct AngleBracketedArgs {
     /// The overall span.
     pub span: Span,
-    /// The arguments for this path segment.
-    pub args: Vec<GenericArg>,
-    /// Constraints on associated types, if any.
-    /// E.g., `Foo<A = Bar, B: Baz>`.
-    pub constraints: Vec<AssocTyConstraint>,
+    /// The comma separated parts in the `<...>`.
+    pub args: Vec<AngleBracketedArg>,
+}
+
+/// Either an argument for a parameter e.g., `'a`, `Vec<u8>`, `0`,
+/// or a constraint on an associated item, e.g., `Item = String` or `Item: Bound`.
+#[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
+pub enum AngleBracketedArg {
+    /// Argument for a generic parameter.
+    Arg(GenericArg),
+    /// Constraint for an associated item.
+    Constraint(AssocTyConstraint),
```

Comment:
At AST/syntax level this should be a variant of `GenericArg`, IMO.
It would make everything simpler and would help to avoid all those "`angle_args`" and the terminology duplication.

Only after lowering to HIR the difference between ty/const/lt args and constraints becomes important, before that they are just components of a single list with different grammars.

Reply:
Putting naming aside, I'm not sure this would simplify things, at least not in terms of parsing (maybe some of the visitors, but that doesn't seem notable). In particular, I would like to have one way of parsing `lifetime | type | anon_const` and then use that to the right of `ident = `. So I think the ideal grammar here would be:
```rust
TermArg = Lifetime | ConstArg | Type ;
GenericArg = TermArg | Ident "=" TermArg | Ident ":" Bounds ;
```
It's true that we don't have associated lifetimes, but supporting `Foo<Bar = CONST>` means we're one variant out. At least in the parser, I would like to have this 3-variant enum equivalent to `TermArg` so that it can be reused for `AssocItem = TermArg` for recovery purposes and to spport https://github.com/rust-lang/rust/issues/70256. That could also be achieved with `unreachable!()` for `::Constraint` in your setup, but that doesn't seem ideal?